### PR TITLE
168 make add product form functional, also issues #217 and #215

### LINKED ISSUE #168: Make Add Product Form Functional **Edit: Just add image**
**Is your feature request related to a problem? Please describe.**
When I am in the product form, and I fill in all the fields, I get an error when saving.

### Acceptance Criteria

- [ ] 'Nombre' and 'Precio' is obligatory
- [ ]  When an image is entered it is displayed
- [ ] When you click save, the information is saved in the database.
- [ ] When you click save, I am redirected to all products.

**Design**
chrome_OqWnktaiS0

### Steps to show

1. Go to 'Proveedores'
2. Click any provider
3. Click Añadir nuevo producto

### Ref for api request and endpoint

- controllers\products.ts
- pages\api\products\index.ts

# Referencias

- Por favor terminar en 3 días. Mas información: https://github.com/Azordev/did-admin-panel/issues/192

### PR DESCRIPTION
**Pull request by PrototypeK02**
### Description

Please include a summary of the change.

El componente funciona de forma correcta con validaciones, acepta imagenes y al crear el producto aparece en la lista con todos los demas productos, acepta imagenes y gif, anteriormente la funcion creaba el producto de forma directa con el graphql lo cual presentaba un error ya que no procesaba las imagenes, se cambio la funcion para que llamase al endpoint correcto.

#### Screenshots

If applicable, add screenshots about your changes

#### Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How to test it

If applicable, describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

### Checklist:

- [ ] Doesn't break the current code.
- [ ] Passes linters and test, also is building.
- [ ] Doesn't have spelling or grammatical problems.
- [ ] Doesn't have unnecessary comments or debugging code.
- [ ] The branch is updated with the last dev branch changes.

#217 #215

### CODE REVIEW

**Review by e-azocar — CHANGES_REQUESTED** [review]
Please check this out
image

**Comment by Eghinner** [comment]
It does not seem to work when adding an image

**Review by heliomar-pena — APPROVED** [review]
Tested and is working properly. Please fix typescript and eslint errors 
imagen
lastrecord

File: views/Products/Create/Form.tsx [hunk 1]
```diff
@@ -1,16 +1,39 @@
-import { FC } from 'react'
-import { useForm } from 'react-hook-form'
+import Image from 'next/image'
+import { ChangeEvent, FC, useState } from 'react'
+import { SubmitHandler, useForm } from 'react-hook-form'
 
 import { MutableProductFormProps, ProductEditable } from '@/services/GraphQL/products/types'
+import stylesInput from '@/styles/EditEvent.module.scss'
+
+// import styles from './SaveorDelete.module.scss'
```
Comment (heliomar-pena) [inline views/Products/Create/Form.tsx]:
```suggestion
```

File: views/Products/Create/Form.tsx [hunk 2]
```diff
@@ -1,16 +1,39 @@
...
+  const [imageUrl, setImageUrl] = useState<string | null>(null)
   const {
     register,
     handleSubmit,
     reset,
     formState: { errors },
-  } = useForm<ProductEditable>()
-  const submitHandler = handleSubmit(onSubmit)
+  } = useForm<ProductEditableWithImg>()
+  const submitHandler = handleSubmit(onSubmit as unknown as SubmitHandler<ProductEditableWithImg>)
```
Comment (heliomar-pena) [inline views/Products/Create/Form.tsx:27]: Why is needed the double "as"?

Could not be just:

```suggestion
  const submitHandler = handleSubmit(onSubmit as SubmitHandler<ProductEditableWithImg>)
```

?

File: views/Products/Create/Form.tsx [hunk 3]
```diff
@@ -32,25 +57,51 @@ const CreateProductForm: FC<MutableProductFormProps> = ({ onSubmit, loading }) =
...
+            onChange={handleFile}
+          />
+          <label htmlFor="image-file" className={stylesInput.image}>
+            <figure>
+              <Image
+                width={imageUrl ? 300 : 40}
+                height={imageUrl ? 200 : 40}
+                objectFit="contain"
+                src={imageUrl || 'https://img.icons8.com/ios/100/image.png'}
+                alt="Imagen del evento"
```
Comment (heliomar-pena) [inline views/Products/Create/Form.tsx]: Please change this alt for a correct value, for example: "Imagen del producto"
imagen

File: views/Products/Create/Form.tsx [hunk 4]
```diff
@@ -32,25 +57,51 @@ const CreateProductForm: FC<MutableProductFormProps> = ({ onSubmit, loading }) =
...
+              />
+            </figure>
+            <span className={stylesInput.label}>{imageFile?.name ? 'Cambiar imagen' : 'Añadir imagen'}</span>
+          </label>
+        </div>
+      </div>
```
Comment (heliomar-pena) [inline views/Products/Create/Form.tsx]: When I try to edit a product it works properly, but the image can't be changed and is empty when should be the image that I added to the product before
lastrecord

**Review by heliomar-pena — CHANGES_REQUESTED** [review]
Edit product image don't works

File: views/Products/Create/Form.tsx (hunk 2 above)
Comment (e-azocar) [reply views/Products/Create/Form.tsx:27]: `as` only will change the type if `onSubmit` is asignable to `SubmitHandler<ProductEditableWithImg>`, so, I'ts needed change type to `unknown`
[@victor-heliomar check this](https://stackoverflow.com/questions/69399211/typescript-why-does-as-unknown-as-x-work)

Comment (heliomar-pena) [reply views/Products/Create/Form.tsx:27]: Thanks for clarify, will check

**Comment by PrototypeK02** [comment]
Se hicieron los cambios necesarios, se hizo funcionar el edit product ya que no guardaba los cambios ni mostraba imagenes, se colocaron alertas cuando los tamaños de las imagenes exceden los 8mb

File: views/Products/Create/Form.tsx (hunk 3 above)
Comment (Israel-Laguan) [reply views/Products/Create/Form.tsx]: @PrototypeK02 don't forget this fix

File: views/Products/Create/Form.tsx (hunk 1 above)
Comment (Israel-Laguan) [reply views/Products/Create/Form.tsx]: @PrototypeK02 don't forget this fix

File: controllers/products.ts [hunk 5]
```diff
@@ -100,6 +101,7 @@ export const createProduct = (req: NextApiRequest, res: NextApiResponse) => {
 }
 
 export const updateProduct = async (req: NextApiRequest, res: NextApiResponse) => {
+  console.log('hola como estas')
```
Comment (Israel-Laguan) [inline controllers/products.ts]: Remove this console.log

File: package.json [hunk 6]
```diff
@@ -91,7 +91,7 @@
     "react-hook-form": "7.30.0",
     "react-toastify": "8.2.0",
     "sharp": "0.30.7",
-    "sweetalert2": "11.4.32",
+    "sweetalert2": "^11.4.32",
```
Comment (Israel-Laguan) [inline package.json]:
```suggestion
    "sweetalert2": "11.4.32",
```

File: pages/productos/editar/[id].tsx [hunk 7]
```diff
@@ -1,28 +1,41 @@
...
+      await axios.put(`/api/products/${query.id}`, form, {
+        headers: {
+          'Content-Type': 'multipart/form-data',
+        },
+      })
 
-  if (mutationError) logError(mutationError, 'pages/productos/[id].tsx', 'Error al actualizar el producto')
+      push('/productos')
+    } catch (error) {
+      console.log(error)
```
Comment (Israel-Laguan) [inline pages/productos/editar/[id].tsx]:
```suggestion
      console.error(error)
```

File: views/Products/Create/Form.tsx [hunk 8]
```diff
@@ -1,16 +1,43 @@
...
-import { useForm } from 'react-hook-form'
+import Image from 'next/image'
+import { ChangeEvent, FC, useState } from 'react'
+import { SubmitHandler, useForm } from 'react-hook-form'
+import Swal from 'sweetalert2'
 
 import { MutableProductFormProps, ProductEditable } from '@/services/GraphQL/products/types'
+import stylesInput from '@/styles/EditEvent.module.scss'
+
+// import styles from './SaveorDelete.module.scss'
```
Comment (Israel-Laguan) [inline views/Products/Create/Form.tsx]:
```suggestion
```

File: views/Products/Edit/index.tsx [hunk 9]
```diff
@@ -57,10 +83,31 @@ const EditProductForm: FC<MutableProductFormProps> = ({ onSubmit, originalData }
...
+            onChange={handleFile}
+          />
+          <label htmlFor="image-file" className={stylesInput.image}>
+            <figure>
+              <Image
+                width={imageUrl ? 300 : 40}
+                height={imageUrl ? 200 : 40}
+                objectFit="contain"
+                src={imageUrl || 'https://img.icons8.com/ios/100/image.png'}
+                alt="Imagen del evento"
```
Comment (Israel-Laguan) [inline views/Products/Edit/index.tsx:107]: Change this alt, please, for example "Imagen del producto"

File: pages/productos/editar/[id].tsx (hunk 7 above)
Comment (PrototypeK02) [reply pages/productos/editar/[id].tsx]: done

File: views/Products/Create/Form.tsx (hunk 8 above)
Comment (PrototypeK02) [reply views/Products/Create/Form.tsx]: it was done already

File: views/Products/Edit/index.tsx (hunk 9 above)
Comment (PrototypeK02) [reply views/Products/Edit/index.tsx:107]: done

File: package.json (hunk 6 above)
Comment (PrototypeK02) [reply package.json]: done

File: controllers/products.ts (hunk 5 above)
Comment (PrototypeK02) [reply controllers/products.ts]: done

File: views/Products/Create/Form.tsx (hunk 3 above)
Comment (PrototypeK02) [reply views/Products/Create/Form.tsx]: done

**Review by Israel-Laguan — APPROVED** [review]

File: views/Products/Edit/index.tsx (hunk 9 above)
Comment (Israel-Laguan) [reply views/Products/Edit/index.tsx:107]: Thanks a lot! Good job

**Comment by Israel-Laguan** [comment]
@angelik0828 please check

**Comment by angelik0828** [comment]
## ✔️ Once the review was done, it was found that the following actions can be taken:

- I can enter the name of the product
- Price of the product
- Product description
- Upload product image
- Save and delete

image

image

## ⚠️ When performing validations on empty fields, use:

- The fonts of the text field.
- Place the validation text below the text field in red color #F5282F the same delete button and smaller font.
- Also set the border of the text field to red #F5282F.

image

image

**- The description field is not required**


## ❌ is not creating product, I'm not sure, but if it's because of the image size, the error should be below the image field and show that it exceeds the size.

image

## ❌  I created this product but I don't save the image, It also didn't show me that the product was created successfully, I followed the list of products and I didn't know what happened.
 

image


@PrototypeK02 please review and correct, thanks for the help

**Comment by e-azocar** [comment]
@angelik0828 it doesn't show image because of image is format SVG, please try using a PNG or JPG image

**Comment by Israel-Laguan** [comment]
> @angelik0828 it doesn't show image because of image is format SVG

What about adding that validation, both in the form's input and in the controller?

Example in the input

```html
<input type="file" name="myImage" accept="image/x-png,image/gif,image/jpeg" />
```

https://www.w3docs.com/snippets/html/how-to-allow-the-file-input-type-to-accept-only-image-files.html

File: pages/api/products/index.ts [hunk 10]
```diff
@@ -1,6 +1,6 @@
 import type { NextApiRequest, NextApiResponse } from 'next'
 
-import { createProduct, getProductsAPI } from '../../../controllers/products'
+import { createProduct, getProductsAPI, updateProduct } from '../../../controllers/products'
```
Comment (e-azocar) [inline pages/api/products/index.ts]: Please remove this import

```suggestion
import { createProduct, getProductsAPI } from '../../../controllers/products'
```

File: views/Products/Create/Form.tsx [hunk 11]
```diff
@@ -1,16 +1,41 @@
...
     handleSubmit,
     reset,
     formState: { errors },
-  } = useForm<ProductEditable>()
-  const submitHandler = handleSubmit(onSubmit)
+  } = useForm<ProductEditableWithImg>()
+  const submitHandler = handleSubmit(onSubmit as unknown as SubmitHandler<ProductEditableWithImg>)
+  const handleFile = (evt: ChangeEvent<HTMLInputElement>) => {
+    const file = evt.target.files?.[0]
+    if (file && file.size >= 8000000) {
```
Comment (e-azocar) [inline views/Products/Create/Form.tsx]: Please avoid _magic numbers_ (define number in constant)
```suggestion
    if (file && file.size >= MAX_FILE_SIZE) {
```

File: views/Products/Edit/index.tsx [hunk 12]
```diff
@@ -20,6 +29,23 @@ const EditProductForm: FC<MutableProductFormProps> = ({ onSubmit, originalData }
 
   const showModal = () => {}
 
+  const handleFile = (evt: ChangeEvent<HTMLInputElement>) => {
+    const file = evt.target.files?.[0]
+    if (file && file.size >= 8000000) {
```
Comment (e-azocar) [inline views/Products/Edit/index.tsx]: Same here

**Review by e-azocar — APPROVED** [review]
It's ok, only some comments

**Review by heliomar-pena — CHANGES_REQUESTED** [review]
Works properly in my side
imagen

But the edit form is not showing the image of the product
imagen

**Comment by PrototypeK02** [comment]
> Works properly in my side imagen
> 
> But the edit form is not showing the image of the product imagen

This works properly in my side, both, add and edit

**Comment by Israel-Laguan** [comment]
@Eghinner @angelik0828  please check

**Comment by Eghinner** [comment]
> @Eghinner @angelik0828 please check

Works

**Comment by angelik0828** [comment]
I also put the validation for the SVG image type, I am validating the other functions

**Comment by PrototypeK02** [comment]
Sorry guys, pushed some changes to this same pull request, issues 217 and 215 are fixed, and working properly, please let me know if you need something else.

**Review by Israel-Laguan — APPROVED** [review]

**Comment by angelik0828** [comment]
## verified

- Validation is done with text fields.
- The product is being created, the form was validated and the creation in the list was validated

image
image

**Review by angelik0828 — APPROVED** [review]

**Review by PrototypeK02 — COMMENTED** [review]
se agrego reset al form, por algun motivo en mi visual aparece pero no se subio

**Review by PrototypeK02 — COMMENTED** [review]
comma added

**Review by heliomar-pena — APPROVED** [review]
Great work !

**Pull request merged**